src/PowerShellEditorServices/Language/TokenOperations.cs

```diff
@@ -0,0 +1,275 @@
+            Token[] tokens,
+            TokenKind startTokenKind,
+            TokenKind endTokenKind,
+            string matchKind)
+        {
+            List<FoldingReference> result = new List<FoldingReference>();
+            Stack<Token> tokenStack = new Stack<Token>();
+            foreach (Token token in tokens)
+            {
+                if (token.Kind == startTokenKind) {
+                    tokenStack.Push(token);
+                }
+                if ((tokenStack.Count > 0) && (token.Kind == endTokenKind)) {
+                    result.Add(CreateFoldingReference(tokenStack.Pop(), token, matchKind));
+                }
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Given an array of token, finds a specific token
+        /// </summary>
+        static private List<FoldingReference> MatchTokenElement(
+            Token[] tokens,
+            TokenKind tokenKind,
+            string matchKind)
+        {
+            List<FoldingReference> result = new List<FoldingReference>();
+            foreach (Token token in tokens)
+            {
+                if ((token.Kind == tokenKind) && (token.Extent.StartLineNumber != token.Extent.EndLineNumber)) {
+                    result.Add(CreateFoldingReference(token, token, matchKind));
+                }
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Returns true if a Token is a block comment;
+        /// - Must be a TokenKind.comment
+        /// - Must be preceeded by TokenKind.NewLine
+        /// - Token text must start with a '#'.false  This is because comment regions
+        ///   start with '&lt;#' but have the same TokenKind
+        /// </summary>
+        static private bool IsBlockComment(int index, Token[] tokens) {
+            Token thisToken = tokens[index];
+            if (thisToken.Kind != TokenKind.Comment) { return false; }
+            if (index == 0) { return true; }
+            if (tokens[index - 1].Kind != TokenKind.NewLine) { return false; }
+            return thisToken.Text.StartsWith("#");
+        }
+
+        /// <summary>
+        /// Finding blocks of comment tokens is more complicated as the newline characters are not
+        /// classed as comments.  To workaround this we search for valid block comments (See IsBlockCmment)
+        /// and then determine contiguous line numbers from there
+        /// </summary>
+        static private List<FoldingReference> MatchBlockCommentTokenElement(
+            Token[] tokens,
+            string matchKind)
+        {
+            // This regular expressions is used to detect a line comment (as opposed to an inline comment), that is not a region
+            // block directive i.e.
+            // - No text between the beginning of the line and `#`
+            // - Comment does start with region
+            // - Comment does start with endregion
+            string lineCommentText = @"\s*#(?!region\b|endregion\b)";
```

Comment:
Given our perf issues, I would take "a little complicated" if it works significantly faster than the alternative. :-)  Man, I really wish the AST preserved comments/whitespace trivia like the C# AST does.

Reply:
Welllllllllllllll, the PowerShell AST does have `Extent`s, so the comments should all be there.

But, I'm guessing what you really want is an AST element that contains special kinds of comments.

I'm pretty sure it doesn't keep regions, but perhaps it should.

I know an AST stores `#requires`s, and probably stores help comments as well.

Time for a new issue in PowerShell??

Reply:
That issue already exists - vote it up!  https://github.com/PowerShell/PowerShell/issues/1806

Reply:
Also theoretically the compiled regex would be even faster than my code, since I do some string operations, whereas it will just be a state-drive char stream consumer. So for example, where my method looks along the string for `"region"` and then again for `"endregion"`, the regex will have a state that represents seeing either. At least it will if it's a true regex with no lookahead or lookbehind and it can be compiled with [Thompson's construction](https://en.wikipedia.org/wiki/Thompson%27s_construction).

Reply:
>AST element that contains special kinds of comments.
Something I can use with AST's FindAll().

Reply:
I don't think comments in AST is a good idea anyway e.g. If you had mismatched `# regions` that would cause you to throw a "Syntax Error" (much like mismatched brackets).  But as compiler, I don't care about missmatched comments

`# regions` are an editor hint, and not part of the official PowerShell language.

Reply:
The forward thinking approach that the C# compiler took (Roslyn) is that the AST is as much for powering eco-system tools like editors, analyzers, etc as it is for compilation.  Somehow they've figured out how to deal with regions in their trivia - `SyntaxFactory.RegionDirectiveTrivia()`.

Reply:
> and not part of the official PowerShell language.

... but if they DID become part of the PS Language ... go for it.

Reply:
I do wonder if traversing the tokens like I am is just as fast (if not more-so) then generating a full blown AST.  A task for another day.  My code will need a refactor at some point.  I traverse all of the tokens about 9+ times.  I could optimise it to 2, maybe 3, full array scans.

Reply:
But that's for another day, another PR.  Get it working, and then get it fast.

Reply:
Your way is definitely faster -- but we already generate an AST with every code change

Reply:
(For clarity: definitely for another day)